src/main/java/org/apache/maven/shared/artifact/filter/PatternIncludesArtifactFilter.java

```diff
@@ -255,540 +183,319 @@ public void reportMissedCriteria( final Logger logger )
-        char ch;
-
-        boolean containsStar = false;
-        for ( char aPatArr : patArr )
-        {
-            if ( aPatArr == '*' )
-            {
-                containsStar = true;
-                break;
-            }
-        }
-
-        if ( !containsStar )
-        {
-            if ( isVersion && ( patArr[0] == '[' || patArr[0] == '(' ) )
-            {
-                return isVersionIncludedInRange( String.valueOf( strArr ), String.valueOf( patArr ) );
-            }
-            // No '*'s, so we make a shortcut
-            if ( patIdxEnd != strIdxEnd )
-            {
-                return false; // Pattern and string do not have the same size
-            }
-            for ( int i = 0; i <= patIdxEnd; i++ )
-            {
-                ch = patArr[i];
-                if ( ch != '?' && ch != strArr[i] )
-                {
-                    return false; // Character mismatch
-                }
-            }
-            return true; // String matches against pattern
-        }
-
-        if ( patIdxEnd == 0 )
-        {
-            return true; // Pattern contains only '*', which matches anything
-        }
-
-        // Process characters before first star
-        while ( ( ch = patArr[patIdxStart] ) != '*' && strIdxStart <= strIdxEnd )
-        {
-            if ( ch != '?' && ch != strArr[strIdxStart] )
-            {
-                return false; // Character mismatch
-            }
-            patIdxStart++;
-            strIdxStart++;
-        }
-        if ( strIdxStart > strIdxEnd )
+        requireNonNull( artifact );
+        return coordinate ->
         {
-            // All characters in the string are used. Check if only '*'s are
-            // left in the pattern. If so, we succeeded. Otherwise failure.
-            for ( int i = patIdxStart; i <= patIdxEnd; i++ )
+            requireNonNull( coordinate );
+            switch ( coordinate )
             {
-                if ( patArr[i] != '*' )
-                {
-                    return false;
-                }
+                case GROUP_ID:
+                    return artifact.getGroupId();
+                case ARTIFACT_ID:
+                    return artifact.getArtifactId();
+                case BASE_VERSION:
+                    return artifact.getBaseVersion();
+                case CLASSIFIER:
+                    return artifact.hasClassifier() ? artifact.getClassifier() : null;
+                case TYPE:
+                    return artifact.getType();
+                default:
             }
-            return true;
-        }
+            throw new IllegalArgumentException( "unknown coordinate: " + coordinate );
+        };
+    }
 
-        // Process characters after last star
-        while ( ( ch = patArr[patIdxEnd] ) != '*' && strIdxStart <= strIdxEnd )
+    private static Artifactoid adapt( final String depTrailString )
+    {
+        requireNonNull( depTrailString );
+        // G:A:T:C:V
```

Comment:
Please document how the pattern looks like with 4 components

Reply:
You want me to document `org.apache.maven.artifact.Artifact#getDependencyTrail()` method here?

Reply:
No, the code processes two pattern style, no? 4 and 5 component. The comment documents on 5 components.

Reply:
https://github.com/apache/maven/blob/master/maven-artifact/src/main/java/org/apache/maven/artifact/DefaultArtifact.java#L224-L248

So it is `G:A:T:C:V` or `G:A:T:V`

Reply:
That's be fine, then we have context.